src/Interpreters/Cache/QueryCache.cpp

```diff
@@ -52,7 +54,40 @@ struct HasNonDeterministicFunctionsMatcher
     }
 };
 
+struct HasSystemTablesMatcher
+{
+    struct Data
+    {
+        const ContextPtr context;
+        bool has_system_tables = false;
+    };
+
+    static bool needChildVisit(const ASTPtr &, const ASTPtr &) { return true; }
+
+    static void visit(const ASTPtr & node, Data & data)
+    {
+        if (data.has_system_tables)
+            return;
+
+        if (const auto * identifier = node->as<ASTTableIdentifier>())
```

Comment:
Probably something like select from `clusterAllReplicas(default, 'system.processes')` will not work here as  'system.query_log' is a `ASTLiteral` or in case ``clusterAllReplicas(default, system.processes)`` it will be `ASTIdentifier`.  Or (I do not know but may be) if the query has `clusterAllReplicas` it is anyway not put into query cache?

Reply:
Good point, thanks! I changed the code a bit so that these cases are now correctly handled.